File: ZSDK_DevDemos_Combined/DeveloperDemosVisualBasic/SmartCardDemoForm.vb
' ********************************************** 
' * CONFIDENTIAL AND PROPRIETARY 
' *
' * The source code and other information contained herein is the confidential and the exclusive property of
' * ZIH Corp. and is subject to the terms and conditions in your end user license agreement.
' * This source code, and any other information contained herein, shall not be copied, reproduced, published, 
' * displayed or distributed, in whole or in part, in any medium, by any means, for any purpose except as
' * expressly permitted under such license agreement.
' * 
' * Copyright ZIH Corp. 2010
' *
' * ALL RIGHTS RESERVED 
' *********************************************** 

Imports System
Imports System.Linq
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports ZSDK_API.Discovery
Imports System.Threading
Imports ZSDK_API.Printer

Namespace ZebraDeveloperDemos
    <System.ComponentModel.DesignerCategory("Code")> _
	Public Partial Class SmartCardDemoForm
		Inherits ZebraForm

		Private Delegate Sub GenericEventHandler()
		Private lastButtonPressed As MenuItem

		Public Sub New()
			AddHandler connectionEstablished, New PrinterConnectionHandler(AddressOf connectionIsEstablished)
			AddHandler connectionClosed, New PrinterConnectionHandler(AddressOf connectionIsClosed)
			InitializeComponent()
			zebraConnectionPanel = connectionPanel
			lastButtonPressed = sendCtAtrButton
		End Sub

		Private Sub ConnectedButtonSettings()
			sendCtAtrButton.Enabled = False
			sendCtDataButton.Enabled = False
		End Sub

		Private Sub connectionIsClosed()
			Invoke(New GenericEventHandler(AddressOf NoConnectionButtonSettings))
		End Sub

		Private Sub NoConnectionButtonSettings()
			sendCtAtrButton.Enabled = True
			sendCtDataButton.Enabled = True
		End Sub

		Private Sub sendCtAtrButtonPressed(sender As Object, e As EventArgs)
			handleButtonPressedEvent(sender)
		End Sub

		Private Sub sendCtDataButtonPressed(sender As Object, e As EventArgs)
			handleButtonPressedEvent(sender)
		End Sub

		Private tempResponseString As [String]

		Private Sub handleButtonPressedEvent(sender As Object)
			Invoke(New GenericEventHandler(AddressOf ConnectedButtonSettings))

			Me.tempResponseString = "Waiting for response..."
			Invoke(New GenericEventHandler(AddressOf updateResponseTextBox))
			lastButtonPressed = DirectCast(sender, MenuItem)
			connect()
		End Sub

		Private Sub updateResponseTextBox()
			responseTextBox.Text = Me.tempResponseString
		End Sub

		Private Sub connectionIsEstablished()
			Dim t As New Thread(AddressOf doCommand)
			t.Start()
			t.Join(10000)
			disconnect()
		End Sub

		Private Sub doCommand()
			Dim reader As SmartcardReader = getPrinter().GetSmartcardReader()
			If reader IsNot Nothing Then
				updateGuiFromWorkerThread("Sending Command...", Color.Gold)
                Thread.Sleep(1000)

				Dim smartCardResponse As Byte()

				If lastButtonPressed Is sendCtAtrButton Then
					smartCardResponse = reader.GetATR()
				Else
					smartCardResponse = reader.DoCommand("8010000008")
				End If
				reader.Close()
				formatAndDisplayResponse(smartCardResponse)
			Else
				updateGuiFromWorkerThread("Printer does not support Smartcard reader", Color.Red)
			End If
		End Sub

		Private Sub formatAndDisplayResponse(response As Byte())
			Dim sb As New StringBuilder()

			For Each b As Byte In response
				sb.Append(b.ToString("X2"))
			Next
			Me.tempResponseString = sb.ToString()
			Invoke(New GenericEventHandler(AddressOf updateResponseTextBox))
		End Sub

		Private Sub SmartCardDemoForm_Closing(sender As Object, e As CancelEventArgs)
			If getConnection() IsNot Nothing Then
				getConnection().Close()
			End If
		End Sub
	End Class
End Namespace

File: ZSDK_DevDemos_Combined/DeveloperDemosCSharp/SmartCardDemoForm.cs
/********************************************** 
 * CONFIDENTIAL AND PROPRIETARY 
 *
 * The source code and other information contained herein is the confidential and the exclusive property of
 * ZIH Corp. and is subject to the terms and conditions in your end user license agreement.
 * This source code, and any other information contained herein, shall not be copied, reproduced, published, 
 * displayed or distributed, in whole or in part, in any medium, by any means, for any purpose except as
 * expressly permitted under such license agreement.
 * 
 * Copyright ZIH Corp. 2010
 *
 * ALL RIGHTS RESERVED 
 ***********************************************/

using System;
using System.Linq;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using ZSDK_API.Discovery;
using System.Threading;
using ZSDK_API.Printer;

namespace ZebraDeveloperDemos {
    public partial class SmartCardDemoForm : ZebraForm {

        private delegate void GenericEventHandler();
        MenuItem lastButtonPressed;

        public SmartCardDemoForm() {
            connectionEstablished += new PrinterConnectionHandler(connectionIsEstablished);
            connectionClosed += new PrinterConnectionHandler(connectionIsClosed);
            InitializeComponent();
            zebraConnectionPanel = connectionPanel;
            lastButtonPressed = sendCtAtrButton;
        }

        private void ConnectedButtonSettings() {
            sendCtAtrButton.Enabled = false;
            sendCtDataButton.Enabled = false;
        }

        private void connectionIsClosed() {
            Invoke(new GenericEventHandler(NoConnectionButtonSettings));
        }

        private void NoConnectionButtonSettings() {
            sendCtAtrButton.Enabled = true;
            sendCtDataButton.Enabled = true;
        }

        private void sendCtAtrButtonPressed(object sender, EventArgs e) {
            handleButtonPressedEvent(sender);
        }
        
        private void sendCtDataButtonPressed(object sender, EventArgs e) {
            handleButtonPressedEvent(sender);
        }

        private String tempResponseString;

        private void handleButtonPressedEvent(object sender) {
            Invoke(new GenericEventHandler(ConnectedButtonSettings));

            this.tempResponseString = "Waiting for response...";
            Invoke(new GenericEventHandler(updateResponseTextBox));
            lastButtonPressed = (MenuItem)sender;
            connect();
        }

        private void updateResponseTextBox() {
            responseTextBox.Text = this.tempResponseString;
        }

        private void connectionIsEstablished() {
            Thread t = new Thread(doCommand);
            t.Start();
            t.Join(10000);
            disconnect();
        }

        private void doCommand() {
            SmartcardReader reader = getPrinter().GetSmartcardReader();
            if (reader != null) {
                updateGuiFromWorkerThread("Sending Command...", Color.Gold);
                Thread.Sleep(1000);

                byte[] smartCardResponse;

                if (lastButtonPressed == sendCtAtrButton) {
                    smartCardResponse = reader.GetATR();
                } else {
                    smartCardResponse = reader.DoCommand("8010000008");
                }
                reader.Close();
                formatAndDisplayResponse(smartCardResponse);
            } else {
                updateGuiFromWorkerThread("Printer does not support Smartcard reader", Color.Red);
            }
        }

        private void formatAndDisplayResponse(byte[] response) {
            StringBuilder sb = new StringBuilder();

            foreach (byte b in response) {
                sb.Append(b.ToString("X2"));
            }
            this.tempResponseString = sb.ToString();
            Invoke(new GenericEventHandler(updateResponseTextBox));
        }

        private void SmartCardDemoForm_Closing(object sender, CancelEventArgs e) {
            if (getConnection() != null) {
                getConnection().Close();
            }
        }
    }
}